doc/developer/design/20220203_connectors.md

```diff
@@ -0,0 +1,211 @@
+sinks.
+
+## Overview
+
+Many users of Materialize create families of sources and sinks that share many
+configuration parameters. The current design of `CREATE SOURCE` and `CREATE
+SINK` make this a verbose affair. The problem is particularly acute with
+Avro-formatted Kafka sources that configure authentication:
+
+```sql
+CREATE SOURCE kafka1
+FROM KAFKA BROKER 'kafka:9092' TOPIC 'top1' WITH (
+    security_protocol = 'SASL_SSL',
+    sasl_mechanisms = 'PLAIN',
+    sasl_username = 'username',
+    sasl_password = 'password',
+)
+FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY 'https://schema-registry' WITH (
+    username = 'username',
+    password = 'password'
+);
+
+CREATE SOURCE kafka2
+FROM KAFKA BROKER 'kafka:9092' TOPIC 'top2' WITH (
+    security_protocol = 'SASL_SSL',
+    sasl_mechanisms = 'PLAIN',
+    sasl_username = 'username',
+    sasl_password = 'password'
+)
+FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY 'https://schema-registry' WITH (
+    username = 'username',
+    password = 'password'
+);
+```
+
+These two source definition differ only in their topic specification, but
+must duplicate eight other configuration parameters.
+
+With the connectors proposed in this document, the `CREATE SOURCE` can instead
+share all the relevant configuration:
+
+```sql
+CREATE CONNECTOR kafka FOR
+KAFKA BROKER 'kafka:9092' WITH (
+    security_protocol = 'SASL_SSL',
+    sasl_mechanisms = 'PLAIN',
+    sasl_username = 'username',
+    sasl_password = 'password'
+);
+
+CREATE CONNECTOR schema_registry FOR
+CONFLUENT SCHEMA REGISTRY 'https://schema-registry' WITH (
+    username = 'username',
+    password = 'password'
+);
+
+CREATE SOURCE kafka1
+FROM KAFKA CONNECTOR kafka TOPIC 'top1'
+FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTOR schema_registry;
+
+CREATE SOURCE kafka2
+FROM KAFKA CONNECTOR kafka TOPIC 'top2'
+FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTOR schema_registry;
+```
+
+## Design
+
+### SQL syntax and semantics
+
+Connector create syntax will come in type specific variants and will not include `with_options`, instead completely specifying the potential options in each variant. 
+
+Using connectors syntactically means replacing the source specific keyword such as BROKER or CONNECTION with CONNECTOR and supplying one or more connector names. 
+
+Properties will be merged by applying values in the supplied order with the last write winning, this supports a usage pattern of supplying base values and then progressively more specific overrides while remaining easy to reason about.  
+
+As long as connectors do not support ALTER then it is strictly not necessary to store the relationship between source and connector in the catalog, instead during planning of the CREATE SOURCE statement the values from the connector can be substituted into relevant places in the `CreateSourceConnector` without requiring any parsing changes for any part of the `CreateSourceStatement`. 
```

Comment:
Are you saying you think any implementation of connectors should be done with a relation from the source to the connector? My current thought had been that for the initial implementation this is unnecessary since the connector is just acting as something to copy values from, so there was no need to take this requirement.

If you think the first version should definitely make explicit the relationship I will have to make larger changes to this plan since I was also actively trying to avoid having to make big changes to the underlying types.

Reply:
Ah, I'm saying that the catalog needs to present the _appearance_ of a dependency between a source and a connector. Specifically, if there are any sources that depend on a connector, then `DROP CONNECTOR` needs to fail unless you use `DROP CONNECTOR CASCADE`. That shouldn't require any major rearchitecting of the connector/source types, just a little bookkeeping in the catalog.